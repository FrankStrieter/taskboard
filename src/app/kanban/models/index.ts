export * from './card';
export * from './task';
export * from './tasks-projection';
