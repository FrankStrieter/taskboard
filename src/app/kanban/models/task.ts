import { Card } from './card';

export interface Task extends Card {
  guid: string;
  isFavorite: boolean;
  isInProgress: boolean;
  isComplete: boolean;
}
