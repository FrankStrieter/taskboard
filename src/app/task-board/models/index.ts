export * from './mode';
export * from './task';
