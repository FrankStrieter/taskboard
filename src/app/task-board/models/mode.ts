export enum Mode {
  ReadOnly,
  Write
}
