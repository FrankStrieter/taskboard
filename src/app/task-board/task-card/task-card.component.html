<h3 name="task-title">{{ task.title }}</h3>
<p name="task-text">{{ task.title }}</p>

<button
  [hidden]="state === mode.ReadOnly"
  (click)="activateReadOnlyMode()"
  name="activate-read-only-mode"
>
  Cancel
</button>
