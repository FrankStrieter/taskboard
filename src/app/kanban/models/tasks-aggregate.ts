import { Task } from './task';

export class TasksAggregate {
  count = this.items.length;

  constructor(public items: Task[] = []) {}
}
