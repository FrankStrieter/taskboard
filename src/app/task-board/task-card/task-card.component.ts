import {
  AfterViewInit,
  Component,
  ElementRef,
  Renderer2,
  Input
} from '@angular/core';
import { Mode } from '../models/mode';

@Component({
  selector: 'tb-task-card',
  templateUrl: './task-card.component.html',
  styleUrls: ['./task-card.component.scss']
})
export class TaskCardComponent implements AfterViewInit {
  @Input() task: { title: string; text: string };

  mode = Mode;
  state = Mode.ReadOnly;

  constructor(
    private _renderer: Renderer2,
    private _host: ElementRef<HTMLElement>
  ) {}

  ngAfterViewInit(): void {
    const remove = this._renderer.listen(
      this._host.nativeElement,
      'click',
      () => this._activateEditModeOnce(remove)
    );
  }

  activateReadOnlyMode() {
    this.state = Mode.ReadOnly;
  }

  activateEditMode() {
    this.state = Mode.Edit;
  }

  private _activateEditModeOnce(remove: () => void): void {
    this.activateEditMode();
    remove();
  }
}
