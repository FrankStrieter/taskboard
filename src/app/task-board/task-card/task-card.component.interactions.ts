import { DebugElement } from '@angular/core';
import { ComponentFixture } from '@angular/core/testing';
import { By } from '@angular/platform-browser';
import { TaskCardComponent } from './task-card.component';

export class TaskCardInteractions {
  constructor(private _fixture: ComponentFixture<TaskCardComponent>) {}

  get nativeElement(): HTMLElement {
    return this._fixture.nativeElement;
  }

  get componentInstance(): TaskCardComponent {
    return this._fixture.componentInstance;
  }

  get debugElement(): DebugElement {
    return this._fixture.debugElement;
  }

  get leaveEditModeButton(): HTMLButtonElement {
    return this._byCss('[name=activate-read-only-mode]') as HTMLButtonElement;
  }

  get viewTemplate(): HTMLElement {
    return this._byCss('tb-task-view');
  }

  get editTemplate(): HTMLElement {
    return this._byCss('tb-task-edit');
  }

  detectChanges(): void {
    this._fixture.detectChanges();
  }

  private _byCss(selector: string) {
    return this._fixture.debugElement.query(By.css(selector)).nativeElement;
  }
}
