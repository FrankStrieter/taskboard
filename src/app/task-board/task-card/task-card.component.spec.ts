import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { Mode } from '../models/mode';
import { TaskCardComponent } from './task-card.component';
import { times } from '@test';
import { By } from '@angular/platform-browser';

describe('<tb-task-card>', () => {
  let sut: ComponentFixture<TaskCardComponent>;
  let leaveEditModeButton: () => HTMLButtonElement;

  beforeEach(async(() => {
    TestBed.configureTestingModule({ declarations: [TaskCardComponent] });
    sut = TestBed.createComponent(TaskCardComponent);

    sut.detectChanges();

    leaveEditModeButton = () =>
      sut.debugElement.query(By.css('[name=activate-read-only-mode]'))
        .nativeElement;
  }));

  describe('When it is created', () => {
    it('should be in read only mode', () => {
      expect(sut.componentInstance.state).toBe(Mode.ReadOnly);
    });

    it('should hide the cancellation button for edit mode', () => {
      expect(leaveEditModeButton().hasAttribute('hidden')).toBe(true);
    });
  });

  describe('When it is clicked', () => {
    it('should switch to edit mode', () => {
      sut.nativeElement.click();
      sut.detectChanges();

      expect(sut.componentInstance.state).toBe(Mode.Edit);
    });

    it('should not accept clicks if edit mode is active', () => {
      const activateEditMode = spyOn(sut.componentInstance, 'activateEditMode');

      times(2, () => {
        sut.nativeElement.click();
        sut.detectChanges();
      });

      expect(activateEditMode).toHaveBeenCalledTimes(1);
    });

    it('should display the cancellation button for edit mode', () => {
      sut.nativeElement.click();
      sut.detectChanges();

      expect(leaveEditModeButton().hasAttribute('hidden')).toBe(false);
    });
  });

  describe('When the user leaves the edit mode', () => {
    it('should switch back to read only mode', () => {
      sut.nativeElement.click();

      leaveEditModeButton().click();
      sut.detectChanges();

      expect(sut.componentInstance.state).toBe(Mode.ReadOnly);
    });

    it('should switch back to read only mode', () => {
      sut.nativeElement.click();

      leaveEditModeButton().click();
      sut.detectChanges();

      expect(leaveEditModeButton().hasAttribute('hidden')).toBe(true);
    });
  });
});
