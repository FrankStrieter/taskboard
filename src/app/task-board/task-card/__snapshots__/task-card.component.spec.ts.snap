// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<tb-task-card> When it is created should display the description of the task 1`] = `
<p
  name="task-text"
>
  Buy Milk
</p>
`;

exports[`<tb-task-card> When it is created should display the title of the task 1`] = `
<h3
  name="task-title"
>
  Buy Milk
</h3>
`;
